[Mobile Payments] Show onboarding before Set up Tap to Pay on iPhone

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #9370
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

If onboarding is incomplete, a merchant will not be able to complete the Set up Tap to Pay on iPhone flow. Up to now, we’ve prevented the flow from being opened until onboarding is complete, relying on the in-line onboarding offered on the Payments menu.

With the addition of the Universal Link to the Set up Tap to Pay on iPhone flow, that’s no longer sufficient, as we don’t want to make a merchant wait on the menu screen before completing the navigation from their tap on the link.

This change removes the block on opening the set up flow, and shows the onboarding flow when required before set up begins.

N.B. We do not use the existing onboarding presenter, as it was not practical to make the flow seamless when combining it with the PaymentSettingsFlow approach.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Using a US store on an iPhone XS or newer, on iOS 16+

1. Disable your payment gateway in wp-admin, or use some other method to cause the store to fail onboarding
2. Open the app
3. Navigate to Menu > Payments
4. Immediately tap the `Set up Tap to Pay on iPhone` button – observe that it is not disabled, even though onboarding is incomplete
5. Observe that you are shown onboarding
6. Enable your payment gateway in wp-admin
7. Tap `Refresh after Activating` in the app
8. Observe that the flow continues as expected

Repeat the test with the store already passing onboarding
Observe that you are either shown the loading screen, or immediately the `Set up Tap to Pay` screen, depending whether the onboarding result has loaded or not.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://user-images.githubusercontent.com/2472348/233349749-36643afb-b53a-4c14-a736-d266718cb809.mp4


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
9370 Show onboarding before TTP set up flow

When onboarding is incomplete, a merchant will not be able to complete the set up flow. Up to now, we’ve prevented the flow from being opened until onboarding is complete, relying on the in-line onboarding offered on the Payments menu.

With the addition of the Universal Link to the Set up Tap to Pay on iPhone flow, that’s no longer sufficient, as we don’t want to make a merchant wait on the menu screen before completing the navigation from their tap on the link.

This change removes the block on opening the set up flow, and shows the onboarding flow when required before set up begins.

joshheald committed:
9370 Set up TTP enabled based on onboarding state

joshheald committed:
9370 Add cancel button to onboarding

joshheald committed:
9370 Track cancel on onboarding in Set up TTP flow

toupper commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/SetUpTapToPayInformationViewModel.swift:
```diff
@@ -6,7 +6,7 @@ final class SetUpTapToPayInformationViewModel: PaymentSettingsFlowPresentedViewM
     var didChangeShouldShow: ((CardReaderSettingsTriState) -> Void)?
     var didUpdate: (() -> Void)?
-    let learnMoreURL: URL
+    private(set) var learnMoreURL: URL
```
Nice one make it private too

toupper commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsViewModel.swift:
```diff
@@ -94,6 +100,35 @@ final class InPersonPaymentsViewModel: ObservableObject {
+        case .completed(_):
+            return .isFalse
+        default:
+            return.isTrue
```
`return.isTrue` I miss a space between `return` and `.isTrue`. I wonder how it compiles 😅

joshheald replied:
Good spot 😊 fixed in 0b07bac

toupper commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsViewModel.swift:
```diff
@@ -94,6 +100,35 @@ final class InPersonPaymentsViewModel: ObservableObject {
+    /// Updates whether the view this viewModel is associated with should be shown or not
+    /// Notifies the viewModel owner if a change occurs via didChangeShouldShow
+    ///
+    private func reevaluateShouldShow(onboardingState: CardPresentPaymentOnboardingState) {
```
Do we have tests for this logic?

joshheald replied:
No, but I'll write some tomorrow

toupper commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/InPersonPaymentsViewModel.swift:
```diff
@@ -94,6 +100,35 @@ final class InPersonPaymentsViewModel: ObservableObject {
+
+        let didChange = newShouldShow != shouldShow
+
+        shouldShow = newShouldShow
```
minor thing, I guess we can move this assignment to the `if didChange` block

joshheald replied:
Yep, makes sense, done in 165364e

toupper reviewed (approved):
LGTM and it tests well. I just added a couple of non-blockers.

joshheald committed:
9370 Fix typo from code review

joshheald committed:
9370 Update flow VMs to set shouldShow on change only

The PaymentSettingsFlowPresentedViewModels all calculate whether their screen should show when `reevaluateShouldShow` is called.

The existing implementations all checked for a change, before calling their `didChange` closure.

However, they updated the stored `shouldShow` regardless of whether it had changed.

We can do this inside the `if` block instead, which doesn’t make any functional difference, but does make the code a little clearer.
